File: 2020/01_report_repair/aoc01.py
"""Report Repair

Advent of Code 2020, day 1
Solution by Geir Arne Hjelle, 2020-12-01
"""
import pathlib
import sys

debug = print if "--debug" in sys.argv else lambda *_: None


def find_adders(numbers, target=2020):
    """Find two numbers that add up to target"""
    lookup = set(numbers)

    for first in numbers:
        if (target - first) in lookup and first != target - first:
            return first, target - first

    return None, None


def main(args):
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        print(f"\n{file_path}:")

        with file_path.open(mode="r") as fid:
            numbers = [int(line) for line in fid]

        # Part 1
        first, second = find_adders(numbers)
        print(f"{first} + {second} = 2020, and their product is {first * second}")

        # Part 2
        for idx, first in enumerate(numbers):
            second, third = find_adders(numbers[idx + 1 :], 2020 - first)
            if second and third:
                print(
                    f"{first} + {second} + {third} = 2020, "
                    f"and their product is {first * second * third}"
                )
                break


if __name__ == "__main__":
    main(sys.argv[1:])

File: 2020/01_report_repair/aoc01.jl
# Report Repair
#
# Advent of Code 2020, day 1
# Solution by Geir Arne Hjelle, 2020-12-01

using Pipe

# Find two numbers that add up to target. Will return all numbers that are part
# of such pairs.
function find_adders(numbers, target=2020)
    lookup = Set(numbers)
    [n for n in numbers if target - n in lookup]
end


# Solve the problem for one file
function solve(filename)
    println("\n$(filename)")

    # Read from file
    input = open(filename) do fid
        @pipe fid |> readlines .|> parse(Int32, _)
    end

    # Part 1
    input |> find_adders |> prod |> println

    # Part 2
    for (idx, first) in enumerate(input)
        pairs = @pipe input[idx + 1:end] |> find_adders(_, 2020 - first)
        if !isempty(pairs)
            [[first]; pairs] |> prod |> println
            break
        end
    end
end


# Run solve on each file
ARGS .|> solve
